### ISSUE TITLE
Styling support for Active and Keyboard Focus states

### ISSUE THREAD

**Issue by Saulis** [vaadin/vaadin-button#4]
Specs by @jouni

### Active

When using the mouse, the `active` state should remain for the duration of the mousedown/touchstart event, and change to the `default` state after the mouseup/touchend event. Note, that the button should not receive the `focus-ring` state at this point if it didn’t have that state before the user activated the button.

When using the keyboard, and activating the button using the Enter or Spacebar key, the `active` state should be enabled for a brief moment (500ms) and then the previous style should be restored (`default` or `focus-ring`).

### Keyboard Focused

This state should only be enabled when the user has moved the focus to the button using the keyboard Tab key.

**~Focus (browser native focus)~**

I don’t think we need to expose this state at all, at least for now. Not required.

### Implementation proposal

The states should be exposed as attributes on the host element of vaadin-button.

```html
<vaadin-button active>…</vaadin-button>
<vaadin-button focus-ring>…</vaadin-button>
```

### Other requirements

- Documentation and code examples
- Needs visual tests to verify the states are triggered correctly
- Should have basic  performance tests to catch regressions in load and startup time
- No user testing needed, as the documentation should be straightforward regarding this and there aren’t competing options for the API

**Comment by jouni** [comment]
Provided initial version of requirements.

**Comment by Saulis** [comment]
Questions:
- What's the use case for having a separate styles for native and keyboard focus? Can you give some examples?
- Could we have the `active` and `keyboard-focus` as independent states? Would be easier to implement since we wouldn't need to store the previous state when changing `active`. Also, that would more enable more styling options, i.e.different active style for mouse and keyboard.

**Comment by jouni** [comment]
Native focus is applied also when clicking the button with a pointer device. But designers rately want that, they want the button to look likeit was before clicking (ie. default styles). But there needs to be a focus indication when navigating the page/app with the keyboard, so tjere needs to be a way to differentiate these cases. See paper-button for reference, it does the exact same thing. 

And yes, of course the active and keyboard-focus states are independent. Not sure why you think otherwise. My explanation was probably convoluted, sorry for that.

**Comment by jouni** [comment]
Also, we are not using the native focus in our themes, that’s why I weite it as a non-requirement.

**Comment by jouni** [comment]
Update the specs a bit, that when activating with a mouse/pointer device, the state `default` state should be restored, not the previous state (possibly `keyboard-focus`).

**Comment by jouni** [comment]
Nicely timed a11ycast which walks through the exact thing about keyboard-focus state, which is actually part of the CSS spec also (`:focus-ring`): https://youtu.be/ilj2P5-5CjI

**Comment by Saulis** [comment]
Notes from grooming:
- Name `focus-ring` instead of `keyboard-focus`

**Comment by Saulis** [comment]
`active` was not implemented yet in #52.
